### ISSUE TITLE
Docker Lumberjack log driver

### ISSUE THREAD

**Issue by andrewvc** [elastic/logstash#3646]
We should write a lumberjack driver for docker. This is the best way to record logs in docker and we should have first-class support. Since logstash-forwarder is written in go, and docker is written in go, this shouldn't be too bad. 

There is an existing syslog driver, but it behaves badly (causes container to exit when it can't connect). Additionally, syslog is not as nice as lumberjack.

[More about docker log drivers](https://docs.docker.com/reference/run/)

**Comment by jordansissel** [comment]
+1

**Comment by andrewvc** [comment]
@jordansissel since you're more familiar with the LSF code, how hard do you think it'd be to rip that code out and turn it into a logging driver for docker? The existing docker logging drivers seem [pretty simple](https://github.com/docker/docker/blob/master/daemon/logger/syslog/syslog.go)

**Comment by andrewvc** [comment]
Looks like GELF and fluentd have added drivers to the master branch of docker as well: https://gist.github.com/lalyos/8e47c73751864c290c39

**Comment by jordansissel** [comment]
As long as we mark this thing as an experiment until we're confident in its behaviors, I'm OK with going forward with this.

I briefly reviewed the fluent logger and it seems like under network faults, [it drops the whole buffer](https://github.com/fluent/fluent-logger-golang/blob/master/fluent/fluent.go#L153-L157). Further, it appears (on code review) that the logger will call [`panic()` if the downstream receiver is down for too long](https://github.com/fluent/fluent-logger-golang/blob/master/fluent/fluent.go#L215) which is probably _not_ good. 

I don't know what the contract for Docker loggers is. I don't know if all loggers in Docker are expected to be nonblocking, but we should find out if that's expected. LSF currently exhibits behaviors that try to provide guaranteed (even if delayed) delivery of logs, and this causes writers to stall. This means that if Docker expects async delivery of logs, then LSF cannot provide this without a middleware that drops logs intentionally during downstream congestion. If we are not async (aka: dropping messages), then LSF stalls due to downstream congestion could cause Docker-run processes to block writing to stdout (or similar) which will likely cause the user's application to become stuck or otherwise behave oddly.

Basically, I don't want us to provide something without knowing how well we can conform to the contract for the Docker logger API.

**Comment by jordansissel** [comment]
A very brief review of https://github.com/docker/docker/pull/13015 shows it uses https://github.com/Graylog2/go-gelf/. This library [appears to only support GELF over UDP](https://github.com/Graylog2/go-gelf/blob/master/gelf/writer.go#L107) which is lossy.

Given it seems both the gelf and fluent support for Docker's logger appear (via code review) to drop logs silently, we definitely need to find out exactly what the contract is for the Docker logger API and see if LSF can even provide it.

I don't want us to contribute things that drop logs - silent dropping and by-default dropping is bad mojo. Logs are precious!

**Comment by andrewvc** [comment]
@jordansissel I'm 100% with you. Truth is, I don't think they have a great contract. If the syslog endpoint is down, for instance, the container will refuse to start (which is absolutely terrifying). I think we have a fair bit of latitude here, in the worst case we could have our own logger even buffer to the local disk under certain conditions.

That's also not good news WRT gelf and UDP. These drivers are all in their infancy it seems.

**Comment by jordansissel** [comment]
I think this is the interface defined in Docker? https://github.com/docker/docker/blob/master/daemon/logger/logger.go#L27-L29

If so, there's no nearby documentation about the expected behavior of the Log() function.

**Comment by jordansissel** [comment]
Maybe the first step could be a blog post about how to point Docker's existing log plugins at Logstash? (syslog, gelf, and fluent all are supported by Logstash)

**Comment by andrewvc** [comment]
@jordansissel that makes sense. I'm not sure if the gelf and fluent formats are yet included in the current docker release, since they aren't in the official docs. Will investigate next week.

**Comment by lennartkoopmann** [comment]
Can confirm that GELF will be in Docker v1.8.

**Comment by jordansissel** [comment]
@lennartkoopmann woot. It's a starting point! <3

**Comment by andrewvc** [comment]
Thanks @lennartkoopmann . It's a bummer that go doesn't support dynamic libs. It makes for some painful release cycles here.

**Comment by suyograo** [comment]
Closing this for https://github.com/elastic/libbeat/issues/37
